In [2]:
import numpy as np
import pickle
import json

## Data

### Data/logs
* data/logs contains the raw data with the games.

Example of the raw data:

{"round_nr": 4, "score": {"A": 3, "B": 3}, "images": {"A": ["person_umbrella/COCO_train2014_000000210266.jpg", "person_umbrella/COCO_train2014_000000354843.jpg", "person_umbrella/COCO_train2014_000000489469.jpg", "person_umbrella/COCO_train2014_000000568064.jpg", "person_umbrella/COCO_train2014_000000438896.jpg", "person_umbrella/COCO_train2014_000000205931.jpg"], "B": ["person_umbrella/COCO_train2014_000000205931.jpg", "person_umbrella/COCO_train2014_000000122731.jpg", "person_umbrella/COCO_train2014_000000036282.jpg", "person_umbrella/COCO_train2014_000000568064.jpg", "person_umbrella/COCO_train2014_000000438896.jpg", "person_umbrella/COCO_train2014_000000354843.jpg"]}, "common": ["person_umbrella/COCO_train2014_000000568064.jpg", "person_umbrella/COCO_train2014_000000354843.jpg", "person_umbrella/COCO_train2014_000000205931.jpg", "person_umbrella/COCO_train2014_000000438896.jpg"], "highlighted": {"A": [true, false, false, true, true, false], "B": [false, false, false, true, true, true]}, "messages": [{"timestamp": "00:04:36.676254", "turn": 0, "speaker": "B", "agent_id": 723, "message": "pink umbrella gorilla cheese"}, {"timestamp": "00:04:55.009862", "turn": 1, "speaker": "A", "agent_id": 444, "message": "uhh...maybe? i have a pink umbrella. yellow-ish background. lots of people"}, {"timestamp": "00:04:56.239490", "turn": 2, "speaker": "B", "agent_id": 723, "message": "lady with open mouth under grey umbrella"}, {"timestamp": "00:05:01.026222", "turn": 3, "speaker": "B", "agent_id": 723, "message": "<selection> <com> person_umbrella/COCO_train2014_000000354843.jpg"}, {"timestamp": "00:05:04.625624", "turn": 3, "speaker": "A", "agent_id": 444, "message": "yes to :-D lady"}, {"timestamp": "00:05:06.995750", "turn": 4, "speaker": "A", "agent_id": 444, "message": "<selection> <com> person_umbrella/COCO_train2014_000000438896.jpg"}, {"timestamp": "00:05:07.283234", "turn": 4, "speaker": "B", "agent_id": 723, "message": "asain smoke poka dot"}, {"timestamp": "00:05:09.286057", "turn": 5, "speaker": "B", "agent_id": 723, "message": "<selection> <com> person_umbrella/COCO_train2014_000000438896.jpg"},
     

# there are 2500 log json files.

In [19]:
with open('../segmentation/data/0_550_1442_22.json') as log_file:
    log = json.load(log_file)
    

In [20]:
log

{'game_id': 0,
 'domain_id': 22,
 'agent_ids': [550, 1442],
 'agent_labels': ['A', 'B'],
 'start_time': '2018-04-13T23:07:43.911138',
 'feedback': {'B': 'col:5<&>partner_u:5<&>self_u:5<&>text:Thankyouitwasfun',
  'A': 'col:5<&>partner_u:5<&>self_u:5<&>text:ThefirstroundgotstuckandIhadtorefreshittogetittogo.Butoverall,itwasagoodfit.'},
 'rounds': [{'round_nr': 1,
   'score': {'A': 3, 'B': 3},
   'images': {'A': ['person_surfboard/COCO_train2014_000000079720.jpg',
     'person_surfboard/COCO_train2014_000000314741.jpg',
     'person_surfboard/COCO_train2014_000000083580.jpg',
     'person_surfboard/COCO_train2014_000000287084.jpg',
     'person_surfboard/COCO_train2014_000000374388.jpg',
     'person_surfboard/COCO_train2014_000000387812.jpg'],
    'B': ['person_surfboard/COCO_train2014_000000387812.jpg',
     'person_surfboard/COCO_train2014_000000572061.jpg',
     'person_surfboard/COCO_train2014_000000515743.jpg',
     'person_surfboard/COCO_train2014_000000314741.jpg',
     'person_s

* <strong> game_id </strong>: 0, 1 , 2 ... >> always 1 game (of 5 rounds) per files

* <strong> domain_id:</strong> 22 ? no idea what this is. Maybe the domain of the images which will be seen in the rounds
* <strong>agent_ids: </strong> [550, 1442]
* <strong>agent_labels: </strong> [A, B]
* <strong>start_time </strong>
* <strong> feedback:</strong> Feedback message from A and B about the played game ...  Thankyouitwasfun...hefirstroundgotstuckandIhadtorefreshittogetittogo.Butoverall,itwasagoodfit
* <strong>round_nr: </strong> 1 to 5
    * <strong>score: </strong> amount of correct images classified by agent A and B >> {'A': 3, 'B': 3} 0 t/m 3
    * <strong>images </strong> The 6 images which are shown to agent A and B
        * <strong>A </strong> [['person_surfboard/COCO_train2014_000000083580.jpg', img2, img3 ... img6]
        * <strong>B </strong> ['person_surfboard/COCO_train2014_000000079720.jpg', img2, img3 .. img6]
        * <strong>common: </strong> ['person_surfboard/COCO_train2014_000000314741.jpg' > amount of common images. could be 0 t/m 6
        * <strong>highlighted A: </strong> {'A': [True, True, False, False, True, False], >> True = highlighted > so agent is asked to tell if image is common or different. Always 3 images are true
        * <strong>highlighted B:  </strong>'B': [False, True, False, True, False, True]},
        * <strong>messages. </strong> These are ALL the messages for a game round. could be 100 messages per round. taking about the images
            * <strong>Time </strong>: '00:05:07.778613'
            *  <strong>turn: </strong> 0, 1, 2 .... 
            *  <strong>speaker: </strong> 'A' or 'B'
            *  <strong>agent_id:  </strong>550 ... 1442
            *  <strong>message:  </strong>'YOu want to go first?' ... '3 children swimming out in the ocean'... 'selection' 'com' person_surfboard/COCO_train2014_000000314741.jpg'}, ... .<next_round>' ... 'No, i do not have that one'}
            
* <strong>round_nr x : </strong> 
* <strong>round_nr: y</strong> 


### Data splits

Splits the data into train, test, val and dev set. Don't know what the numbers correspond to.

In [4]:
with open('../segmentation/data/data_splits.json') as json_file:
    data_splits = json.load(json_file)

In [6]:
list(data_splits)

['train', 'test', 'val', 'dev']

In [13]:
data_splits['val']

[1106,
 2334,
 1135,
 600,
 2346,
 2330,
 865,
 768,
 934,
 2329,
 661,
 174,
 330,
 1175,
 2020,
 1902,
 230,
 1779,
 1514,
 801,
 644,
 1990,
 518,
 848,
 520,
 1839,
 670,
 43,
 991,
 1495,
 989,
 1849,
 2177,
 1215,
 1213,
 889,
 1570,
 1624,
 1980,
 888,
 1452,
 1841,
 1111,
 772,
 277,
 2323,
 915,
 1653,
 1126,
 917,
 1775,
 421,
 1592,
 345,
 912,
 1650,
 1501,
 1162,
 1651,
 1499,
 1082,
 1521,
 1041,
 1997,
 1086,
 231,
 1412,
 2457,
 1188,
 1882,
 1707,
 907,
 533,
 282,
 1954,
 364,
 1821,
 280,
 1169,
 457,
 1074,
 236,
 828,
 640,
 1770,
 1754,
 1727,
 735,
 2395,
 717,
 1250,
 1527,
 274,
 2060,
 209,
 959,
 2252,
 2234,
 784,
 792,
 107,
 2321,
 1917,
 1382,
 1482,
 1623,
 1336,
 54,
 487,
 562,
 538,
 118,
 858,
 1380,
 561,
 1817,
 754,
 1448,
 1955,
 2491,
 1868,
 1243,
 1694,
 1447,
 639,
 136,
 2065,
 2343,
 2214,
 584,
 587,
 582,
 1389,
 1093,
 204,
 1026,
 2136,
 1901,
 821,
 1446,
 1422,
 1218,
 475,
 1574,
 84,
 526,
 1573,
 1907,
 12,
 1042,
 445,
 1003,
 802

Looking at the test data, we have chains, ids, sections and segments (just as the training data has)

### test_sections.pickle >> 374 sections

In [56]:
sections = pickle.load( open( "../segmentation/data/test_sections.pickle", "rb" ))

sections[0]

(933,
 [([([<utils.processor.Message at 0x12c638c2808>,
     {'person_car/COCO_train2014_000000126097.jpg',
      'person_car/COCO_train2014_000000258505.jpg'}),
    ([<utils.processor.Message at 0x12c638c2ec8>,
     {'person_car/COCO_train2014_000000425647.jpg',
      'person_car/COCO_train2014_000000440460.jpg'})],
   {'person_car/COCO_train2014_000000011619.jpg',
    'person_car/COCO_train2014_000000126097.jpg',
    'person_car/COCO_train2014_000000258505.jpg',
    'person_car/COCO_train2014_000000282514.jpg',
    'person_car/COCO_train2014_000000394151.jpg',
    'person_car/COCO_train2014_000000425647.jpg',
    'person_car/COCO_train2014_000000440460.jpg',
    'person_car/COCO_train2014_000000563545.jpg'}),
  ([([<utils.processor.Message at 0x12c638c2108>,
     {'person_car/COCO_train2014_000000195449.jpg'}),
    ([<utils.processor.Message at 0x12c638c2048>,
     {'person_car/COCO_train2014_000000258505.jpg'}),
    ([<utils.processor.Message at 0x12c638c99c8>,
     {'person_car/COC

In [24]:
pickle.dump(data_splits, open( "save.p", "wb" ) )
data_splits = pickle.load( open( "save.p", "rb" ) )
data_splits

{'train': [2332,
  1259,
  599,
  2335,
  1363,
  741,
  740,
  2209,
  24,
  1836,
  2116,
  2331,
  288,
  2120,
  2208,
  1376,
  1241,
  1029,
  1048,
  333,
  1127,
  815,
  2328,
  977,
  2410,
  2348,
  2293,
  1310,
  2326,
  1304,
  1163,
  1999,
  2411,
  1255,
  612,
  576,
  2068,
  2325,
  1951,
  1001,
  469,
  2412,
  1845,
  1774,
  2046,
  1077,
  2387,
  1164,
  2362,
  2347,
  173,
  988,
  1159,
  85,
  1831,
  758,
  2106,
  2079,
  883,
  1993,
  337,
  2040,
  1994,
  2104,
  1808,
  878,
  1905,
  2165,
  1877,
  882,
  2138,
  460,
  2105,
  2163,
  623,
  2405,
  1878,
  459,
  881,
  879,
  689,
  743,
  547,
  250,
  134,
  1425,
  1417,
  261,
  1307,
  1400,
  1591,
  1235,
  1805,
  495,
  247,
  803,
  1729,
  517,
  536,
  2028,
  2473,
  2391,
  2392,
  1312,
  2066,
  1143,
  2047,
  514,
  2471,
  2390,
  2244,
  2406,
  339,
  910,
  1314,
  1461,
  2157,
  2162,
  185,
  1257,
  1367,
  1459,
  1631,
  990,
  1294,
  671,
  254,
  669,
  1285,
  94

### Test segments
Its a list of length 6801, where each element is a dictionary.

Dict has four keys: ['segment', 'image_set', 'targets', 'length']

* Segment - ooh or maybe the tokens are the indices from the vocab? [6, 23, 13, 759, 20, 8, 28, 51, 12, 7, 43],
* image set - image IDs in the set that is shown, corresponds to imageid in chains and vectors files. There are around 8, 9 or 10 images, possibly the unique images. ['93469', '341060',  '483794',   '259745', '524866','14238','96757','380128'],
* targets - the index of the target image, probably index of the image set. Sometimes there are multiple targets [3, 5, 2],
* length - length of the segment, i.e. number of utterances? ... one length per segment 11, 30




In [6]:
with open('../segmentation/data/test_segments.json') as json_file:
    data_seg = json.load(json_file)

In [7]:
print('number of datapoints:', len(data_seg))

number of datapoints: 6801


In [9]:
data_seg[64]

{'segment': [6, 130, 70, 8, 27, 18, 8, 41, 280, 7, 287],
 'image_set': ['206968',
  '553308',
  '184319',
  '200221',
  '113123',
  '86922',
  '538682',
  '175391'],
 'targets': [6],
 'length': 11}

In [60]:
# Targets show the indices of the targets, 
targets = []
for i in range(10):
    targets.append(data_seg[i]['targets'])
targets

[[5], [0], [6, 2, 3], [1], [4], [2], [7], [6], [4, 6], [8]]

In [61]:
# Targets show the indices of the targets, 
segments = []
for i in range(10):
    segments.append(len(data_seg[i]['image_set']))
segments

[8, 8, 8, 10, 10, 10, 10, 10, 9, 9]

### Test chains
Contains 2811 datapoints

Each datapoints is a dict with ['game_id', 'chain_id', 'target', 'segments', 'lengths']

* Game_id
* chain_id - id for this chain of utterances in this game
* target - target image id, that corresponds with the id in segments and vectors
* segments - a list of indices that correspond to the indices of the test_segments.json file
* lengths - lengths of the separate segments

In [3]:
with open('../segmentation/data/test_chains.json') as json_file:
    data_chain = json.load(json_file)

In [4]:
len(data_chain)

2811

In [5]:
data_chain[642]

{'game_id': '3HHRAGRYX93VK4PLSLP2RF6THARO923U4J9857OF97NRCJ9JMW7ZFOFKG7BZ',
 'chain_id': 3921,
 'target': '348269',
 'segments': [1549, 1562, 1563],
 'lengths': [19, 24, 2]}

Segments correspond to the ids in the test_segment file, looking at the lenght. 341060 is one of the images in data_seg[0]

In [72]:
data_seg[0]['length'], data_seg[7]['length'], data_seg[8]['length']

(35, 15, 33)

### Test ids

A list of game_ids. Don't know if the index corresponds to something..

In [105]:
with open('data/test_ids.json') as json_file:
    data_ids = json.load(json_file)

In [106]:
len(data_ids)

374

In [109]:
data_ids[:5]

['3Z7EFSHGNACKCBAG5WMPCIYUZO6XCI39JEC7537VZS8M5L5HL7OX19JDYVCD',
 '3RWE2M8QWI8V6P1WRA6WJF5VNDKN0Z33JKGHPFYDSS9383GFTUHCC8QGLNMA',
 '3A1PQ49WVIF33I03KVFJS0R96RAH1C3018Q3ZVOJOCM0R6CJK6H5MIXUERA7',
 '3SB5N7Y3O42PWXKA46T1TVWYJIZ0G031EUONYN2W1THB2OVMUYJXVQ400OV0',
 '39PAAFCODNY9U37WR3GXD71D47STV13DY46V3X3QGQB7HRKGGM6T10BG855J']

### Vectors file
containing image feature vectors.

vectors is a dict with 358 entries, with as key the image id. Each image is a featurevector of 2048 dimension

In [73]:
with open('data/vectors.json') as json_file:
    vectors = json.load(json_file)

In [89]:
print('number of images:', len(vectors))

number of images: 358


In [90]:
print('Image keys',list(vectors)[:10])

Image keys ['34815', '86189', '98596', '256981', '266311', '293658', '326202', '386326', '386722', '441488']


In [95]:
print('length of the feature vector for an image:', len(vectors['341060']))

length of the feature vector for an image: 2048


In [94]:
print('part of the feature vec')
vectors['341060'][:15]

part of the feature vec


[0.5908005833625793,
 0.0636805072426796,
 0.26698243618011475,
 0.16566187143325806,
 0.29381296038627625,
 0.2509632706642151,
 0.013470017351210117,
 0.08392883837223053,
 0.10088781267404556,
 0.036955565214157104,
 0.7541999816894531,
 0.319126158952713,
 0.3404830992221832,
 0.17770926654338837,
 0.5060359835624695]

### Vocab
vocab.csv is the vocab, with the names and probably number of times the word occurs

In [90]:
with open('../segmentation/data/vocab.csv') as file:
    counter = 0
    for line in file:
        print(line)
        counter += 1
        if counter == 10:
            break

a,29866

have,28480

i,25617

the,24959

?,20610

with,19288

do,14700

no,14698

.,14624

in,13805



In [91]:
vocab = []
vocab_inv = {}
with open('../segmentation/data/vocab.csv') as file:
    counter = 0
    for line in file:
        line_clean = line.strip().split(',')
        vocab.append(line_clean)
        if line_clean[1] in vocab_inv:
            vocab_inv[line_clean[1]].append(line_clean[0])
        else:
            vocab_inv[line_clean[1]] = [line_clean[0]]
vocab

[['a', '29866'],
 ['have', '28480'],
 ['i', '25617'],
 ['the', '24959'],
 ['?', '20610'],
 ['with', '19288'],
 ['do', '14700'],
 ['no', '14698'],
 ['.', '14624'],
 ['in', '13805'],
 ['on', '13472'],
 ['yes', '12036'],
 ['and', '11763'],
 ['|', '|', '11043'],
 ['that', '9385'],
 ['one', '9341'],
 ['you', '8964'],
 ['of', '7539'],
 ['guy', '6596'],
 ['man', '6255'],
 ['white', '5954'],
 ['red', '5878'],
 ['shirt', '5254'],
 ['to', '5146'],
 ['it', '5048'],
 ["n't", '4632'],
 ['is', '4501'],
 ['blue', '3936'],
 ['black', '3898'],
 ['woman', '3843'],
 ['not', '3439'],
 ['again', '3429'],
 ['two', '3198'],
 ['green', '2892'],
 ['girl', '2665'],
 ['nope', '2477'],
 ['front', '2411'],
 ['bike', '2408'],
 ['ok', '2397'],
 ['cake', '2345'],
 ['lady', '2279'],
 ['orange', '2228'],
 ['holding', '2172'],
 ['laptop', '2091'],
 ['truck', '2061'],
 ['next', '1969'],
 ['!', '1914'],
 ['dog', '1890'],
 ['at', '1833'],
 ['yellow', '1830'],
 ["'s", '1814'],
 ['too', '1729'],
 ['bus', '1695'],
 ['pink', '

In [149]:
# Nope, inverted vocab doesnt work
vocab_inv['212']

KeyError: '212'

In [152]:
for voc_id in data_seg[0]['segment']:
    print(vocab[voc_id], voc_id)

['with', '19298'] 5
['lol', '385'] 212
['?', '20633'] 4
['down', '509'] 166
['with', '19298'] 5
['and', '11763'] 12
['shirt', '5254'] 22
['recliner', '49'] 782
['woman', '3844'] 29
['tank', '335'] 234
['good', '385'] 213
['on', '13482'] 10
['?', '20633'] 4
['bag', '147'] 416
['that', '9390'] 14
['and', '11763'] 12
['shirt', '5254'] 22
['no', '14704'] 7
['do', '14706'] 6
['yellow', '1830'] 49
['yes', '12045'] 11
['do', '14706'] 6
[')', '555'] 155
['statue', '250'] 295
['yes', '12045'] 11
['do', '14706'] 6
['pick', '45'] 811
['you', '8978'] 16
['in', '13812'] 9
['elephant', '970'] 95
['on', '13482'] 10
['with', '19298'] 5
['of', '7545'] 17
['.', '14641'] 8
['and', '11763'] 12


## Models
The training scripts also have code for analysis of the models using the test data.
TODO: See how this works, cause I think we only really need to change our test dataset